packages/compass-components/src/hooks/use-copy-paste-context-menu.tsx

```diff
@@ -0,0 +1,218 @@
+import React, { useCallback, useState, useEffect } from 'react';
+
+import { useContextMenuGroups } from '../components/context-menu';
+
+const NON_TEXT_INPUT_TYPES = [
+  'checkbox',
+  'radio',
+  'submit',
+  'button',
+  'image',
+  'file',
+  'color',
+  'range',
+];
+const SELECTION_EVENTS = [
+  'selectionchange',
+  'focusin',
+  'focusout',
+  'input',
+  'keyup',
+  'mouseup',
+  'contextmenu',
+];
+
+const isEditableTextElement = (
+  element: Element | null
+): element is HTMLElement => {
+  if (!element) return false;
+
+  const tagName = element.tagName.toLowerCase();
+
+  if (tagName === 'input') {
+    const inputType = (element as HTMLInputElement).type.toLowerCase();
+    return (
+      !NON_TEXT_INPUT_TYPES.includes(inputType) &&
+      !(element as HTMLInputElement).readOnly
+    );
+  }
+
+  if (tagName === 'textarea') {
+    return !(element as HTMLTextAreaElement).readOnly;
+  }
+
+  return element.getAttribute('contenteditable') === 'true';
+};
+
+function elementIsTextInput(
+  element: Element | null
+): element is HTMLInputElement | HTMLTextAreaElement {
+  return (
+    !!element && (element.tagName === 'INPUT' || element.tagName === 'TEXTAREA')
+  );
+}
+
+const getSelectionState = () => {
+  const activeElement = document.activeElement;
+  const selection = window.getSelection();
+
+  let selectionStart = 0;
+  let selectionEnd = 0;
+  let selectedText = '';
+
+  if (elementIsTextInput(activeElement)) {
+    selectionStart = activeElement.selectionStart || 0;
+    selectionEnd = activeElement.selectionEnd || 0;
+    selectedText = activeElement.value.substring(selectionStart, selectionEnd);
+  } else if (selection) {
+    selectedText = selection.toString();
+  }
+
+  return {
+    activeElement,
+    selectionStart,
+    selectionEnd,
+    selectedText,
+  };
+};
+
+export function useCopyPasteContextMenu() {
+  const [editCapabilities, setEditCapabilities] = useState({
+    canCut: false,
+    canCopy: false,
+    canPaste: false,
+  });
+  const [contextState, setContextState] = useState<{
+    activeElement: Element | null;
+    selectionStart: number;
+    selectionEnd: number;
+    selectedText: string;
+  } | null>(null);
+
+  useEffect(() => {
+    const captureSelectionState = () => {
+      const selectionState = getSelectionState();
+
+      setContextState(selectionState);
+
+      const hasSelection = selectionState.selectedText.length > 0;
+      const isEditable = isEditableTextElement(selectionState.activeElement);
+
+      setEditCapabilities({
+        canCopy: hasSelection,
+        canCut: hasSelection && isEditable,
+        canPaste: isEditable,
+      });
+    };
+
+    // Listen to events that can change selection or focus.
+    // These impact if and what we can cut/copy/paste.
+    SELECTION_EVENTS.forEach((event) =>
+      document.addEventListener(event, captureSelectionState)
+    );
```

Comment:
Nit: Any reason you prefer this over a for...of loop? I know the language specification guarantees the callback is invoked synchronously, the for...of loop however makes that more apparent IMO and it handles edge-cases with promises a bit more gracefully (i.e. if `addEventListener` returned a promise, that would be "shollowed" and not awaited here). Not feeling strongly here, but wanted to mention and I'm curious if there's a reason for the preference to use `forEach` here.

Reply:
Nope! I also prefer for...of loops! I was likely copy pasting from somewhere or the ai autocompleted and I didn't catch it. Thanks for mentioning.